### PR TITLE
Add eval construct / wire up to Simala's #eval decl; fix / defer some Langium grammar issues; fix some backend parser bugs

### LINKED ISSUE #19: Wire it up to Well-Typed's evaluator for concrete evaluation
- [x] desugar / transform the AST into the AST that Well-Typed's evaluator uses (but without worrying about the entrypoint) -- PR #69 
- [ ] Add surface syntax for entrypoint 

tagging @kosmikus and @fendor so they know this is on my radar (assuming this mini-project doesn't get discarded / veto'd)

### LINKED ISSUE #80: [ToSimala] Look into issue / bug with translation
Edit: Ah, OK, what I need to look into is 

```
> rec x = {}
TypeError TFun TRecord
```
It's probably that I'm marking it as recursive when it shouldn't be marked as such

----------------

```
❯ cabal run lam4-cli -- examples/life_assured_predicate_fun_decl.l4
"------- CST -------------"
[ TypeDecl
    "Accident"
    _1
    (RecordDecl
       []
       []
       RecordDeclMetadata
         { transparency = Transparent , description = Nothing })
, TypeDecl
    "LifeAssured"
    _2
    (RecordDecl
       []
       []
       RecordDeclMetadata
         { transparency = Transparent , description = Nothing })
, Rec
    "payout"
    _3
    (Fun
       RuleMetadata
         { originalRuleRef = Nothing
         , transparency = Transparent
         , description =
             Just "Calculates how much the life assured can get for their claim"
         }
       [ "life_assured" _4 , "accident" _5 ]
       (Lit (IntLit 0)))
, Rec
    "is eligible for a payout"
    _6
    (Predicate
       RuleMetadata
         { originalRuleRef = Nothing
         , transparency = Transparent
         , description =
             Just "Checks if the life assured is eligible for a payout."
         }
       [ "x" _7 ]
       (Lit (BoolLit True)))
, Rec "x" _8 (Record [])
, Eval (PredApp (Var "is eligible for a payout" _6) [ Var "x" _8 ])
]
"-------- Simala exprs ---------"
rec payout_3 = fun (life_assured_4,accident_5) => 0
rec `is eligible for a payout_6` = fun (x_7) => true
rec opaque x_8 = {}
#eval `is eligible for a payout_6`(x_8)
"-------------------------------"
TypeError TFun TRecord
```

Also
```
❯ cabal run lam4-cli -- examples/simple_join_or_record_field_deref.l4
"------- CST -------------"
[ TypeDecl
    "Applicant"
    _1
    (RecordDecl
       [ MkRowTypeDecl
           { name = "publications" _2
           , typeAnnot = BuiltinType BuiltinTypeInteger
           , metadata = Nothing
           }
       ]
       []
       RecordDeclMetadata
         { transparency = Transparent , description = Nothing })
, Rec
    "someApplicant"
    _3
    (Record [ ( "publications" _2 , Lit (IntLit 1000) ) ])
, Rec
    "numberOfPublications"
    _4
    (Fun
       RuleMetadata
         { originalRuleRef = Nothing
         , transparency = Transparent
         , description = Nothing
         }
       [ "applicant" _5 ]
       (Project (Var "applicant" _5) "publications" _2))
, Rec
    "eligible for grant"
    _6
    (Predicate
       RuleMetadata
         { originalRuleRef = Nothing
         , transparency = Transparent
         , description = Nothing
         }
       [ "x" _7 ]
       (BinExpr
          Gt (Project (Var "x" _7) "publications" _2) (Lit (IntLit 10))))
, Rec
    "happy"
    _8
    (Predicate
       RuleMetadata
         { originalRuleRef = Nothing
         , transparency = Transparent
         , description = Nothing
         }
       [ "x" _9 ]
       (PredApp (Var "eligible for grant" _6) [ Var "x" _9 ]))
]
"-------- Simala exprs ---------"
rec opaque someApplicant_3 = {publications_2 = 1000}
rec numberOfPublications_4 = fun (applicant_5) => applicant_5.publications_2
rec `eligible for grant_6` = fun (x_7) => x_7.publications_2 > 10
rec happy_8 = fun (x_9) => `eligible for grant_6`(x_9)
"-------------------------------"
TypeError TFun TRecord
```

### PR DESCRIPTION
**Pull request by ym-han**
1. Closes #19 , #71 , #76 . Surface syntax for #eval is `@REPORT`; happy to change this if people have other ideas.
2. Also enables function application for nullary functions -- grammar had previously unintentionally prevented that
3. Fixes #80 
4. Fixes unary op backend parser bug

Note
* There is a naming / scoping thing I'll need to check / think through (#79 ); this shouldn't be hard
* Need to add tests; haven't tested very much yet

### CODE REVIEW

File: lam4-cli/app/Main.hs [hunk 1]
```diff
@@ -16,7 +16,7 @@ import qualified Lam4.Expr.ConcreteSyntax      as CST (Decl)
 import           Lam4.Expr.Parser              (parseProgramByteStr)
 import           Lam4.Expr.ToConcreteEvalAST   (cstProgramToConEvalProgram)
 import           Lam4.Expr.ToSimala            (SimalaProgram)
-import qualified Lam4.Expr.ToSimala            as ToSimala (compile, render)
+import qualified Lam4.Expr.ToSimala            as ToSimala (compile, render, doEvalDeclsTracing, TraceMode(..), emptyEnv)
```
Comment (fendor) [inline lam4-cli/app/Main.hs:19]:
```suggestion
import qualified Lam4.Expr.ToSimala            as ToSimala
```

Comment (fendor) [reply lam4-cli/app/Main.hs:19]: A little bit simpler I believe

**Review by fendor — APPROVED** [review]
LGTM!

**Comment by ym-han** [comment]
> LGTM!

Thank you for looking through this!

**Pull request merged**